### ISSUE TITLE
Genma Onimusha:res scale does not work for 3D objects (because of downscaling performed in 2D blit/texturing in final draw)

### ISSUE THREAD

**Issue by ko81e24wy** [xemu-project/xemu#1032]
### Title

https://xemu.app/titles/43430001/#Genma-Onimusha

### Bug Description

res scale takes no effect on this game,its a 3D object + 2D background game.
1X
![1X](https://user-images.githubusercontent.com/38323456/173443522-00c42c64-0909-41fb-a71e-f7f358c22206.jpg)
4X
![xemu-2022-06-14-04-42-01](https://user-images.githubusercontent.com/38323456/173443570-b09d8efe-d797-4044-88d5-0129cc6526ab.jpg)
4X on CXBX-R  with depth issue
![ONI-CXBX](https://user-images.githubusercontent.com/38323456/173443645-979fe958-6a5c-4c9e-a0dd-b9339d944a33.jpg)

### Expected Behavior

res scale should work as other game.

### xemu Version

Version:      0.7.28
Branch:       master
Commit:       176b574403e0f4a08bfb1f8ab58e0f819a45cf81
Date:         Fri Jun  3 21:40:57 UTC 2022

### System Information

CPU:          Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
OS Platform:  Windows
OS Version:   Windows
Manufacturer: NVIDIA Corporation
GPU Model:    NVIDIA GeForce GTX 1080/PCIe/SSE2
Driver:       4.0.0 NVIDIA 512.59
Shader:       4.00 NVIDIA via Cg compiler

**Comment by abaire** [comment]
If the background is truly pre-rendered 2d the internal resolution scaling would not be expected to have any effect on it as there is no additional information that can be regained by upscaling. You'd need either a remastered texture at a higher resolution or to use something like AI super resolution to increase.

Are you saying that the scaling has no effect on the 3d objects either? Can you add some screenshots at different scales for comparison?

**Comment by ko81e24wy** [comment]
> If the background is truly pre-rendered 2d the internal resolution scaling would not be expected to have any effect on it as there is no additional information that can be regained by upscaling. You'd need either a remastered texture at a higher resolution or to use something like AI super resolution to increase.
> 
> Are you saying that the scaling has no effect on the 3d objects either? Can you add some screenshots at different scales for comparison?

Update,seems the scaling take effect on the texture,but the aliasing still there.

**Comment by abaire** [comment]
Thanks, unless I'm missing something, this is what I'd expect. There's currently no way to upscale textures to higher resolutions, so it just takes the original texture and stretches it.

I'd suggest changing this from a bug to a feature request to investigate super-resolution image enhancement for textures.

**Comment by ko81e24wy** [comment]
> 

I think texture fliter is good enough.

**Comment by abaire** [comment]
> > 
> 
> I think texture fliter is good enough.

Filters are already specified by the game and blindly overriding mag filters doesn't strike me as a good idea. There will be cases where the filter selection is important (e.g., if the texture is being used to modulate something else and not actually for display). You can look at the pgraph trace output and see what filter this game is using, it's possible that it's trying to use something that isn't implemented, but it's quite possibly just an issue of the texture being magnified many times larger than it was ever intended to be.

**Comment by ko81e24wy** [comment]
> > > 
> > 
> > 
> > I think texture fliter is good enough.
> 
> Filters are already specified by the game and blindly overriding mag filters doesn't strike me as a good idea. There will be cases where the filter selection is important (e.g., if the texture is being used to modulate something else and not actually for display). You can look at the pgraph trace output and see what filter this game is using, it's possible that it's trying to use something that isn't implemented, but it's quite possibly just an issue of the texture being magnified many times larger than it was ever intended to be.

emmm,its a little hard for me to understand...
Do you mean close this issue and open a new a feature request?

**Comment by abaire** [comment]
> > > > 
> > > 
> > > 
> > > I think texture fliter is good enough.
> > 
> > 
> > Filters are already specified by the game and blindly overriding mag filters doesn't strike me as a good idea. There will be cases where the filter selection is important (e.g., if the texture is being used to modulate something else and not actually for display). You can look at the pgraph trace output and see what filter this game is using, it's possible that it's trying to use something that isn't implemented, but it's quite possibly just an issue of the texture being magnified many times larger than it was ever intended to be.
> 
> emmm,its a little hard for me to understand... Do you mean close this issue and open a new a feature request?

No, you can keep this open, please just change the title to "Genma Onimusha: 2d backgrounds are not improved by increasing res scale." so it's clear that only the backgrounds are problematic.

Then @mborgerson can change the label from "bug" to "enhancement" as I don't think anybody else can modify labels.

**Comment by ko81e24wy** [comment]
> > > > > 
> > > > 
> > > > 
> > > > I think texture fliter is good enough.
> > > 
> > > 
> > > Filters are already specified by the game and blindly overriding mag filters doesn't strike me as a good idea. There will be cases where the filter selection is important (e.g., if the texture is being used to modulate something else and not actually for display). You can look at the pgraph trace output and see what filter this game is using, it's possible that it's trying to use something that isn't implemented, but it's quite possibly just an issue of the texture being magnified many times larger than it was ever intended to be.
> > 
> > 
> > emmm,its a little hard for me to understand... Do you mean close this issue and open a new a feature request?
> 
> No, you can keep this open, please just change the title to "Genma Onimusha: 2d backgrounds are not improved by increasing res scale." so it's clear that only the backgrounds are problematic.
> 
> Then @mborgerson can change the label from "bug" to "enhancement" as I don't think anybody else can modify labels.

Gotcha，but the issue is still there,I mean res scale takes no effect on the 3D model in this game.Seems we are talking about different things.(⊙﹏⊙)

**Comment by gordonfreeman01** [comment]
Is this even a bug? The 3D model does look upscaled based on the screenshots and as per earlier discussions the 2D background is likely from source. Some sort of user-specified override filters may help but does not look like a bug to me.

**Comment by SeongGino** [comment]
Looking at the 4X scaled Xemu screenshot specifically, the player model indeed does not look like it's upscaled at all - it's still aliased as if it's internally running at 480p - hence why it just looks like the first (native Xbox res) screenshot blown up.

The CXBX-Reloaded screenshot is purely for comparison of what an internal resolution uplift should look like... or, so I imagine.

Just to make the case clearer: this is how they look on my end with a 4X scale, as indicated in the associated screencap of Xemu's graphics settings
![2023_11-27 233654](https://github.com/xemu-project/xemu/assets/7321839/189a13eb-bb6f-498f-8f8a-c915972df997)
![2023_11-27 233652](https://github.com/xemu-project/xemu/assets/7321839/1045e825-4440-43bd-9c33-9ff9596785c9)

**The HUD and backgrounds are 2D and will look like that regardless;** it's just the character models themselves (player, enemies, NPCs) that aren't as sharp as they should be in another fully 3D game.
![2023_11-27 234017](https://github.com/xemu-project/xemu/assets/7321839/62f3073f-57a1-466c-b489-020f7c9ec3de)

**Comment by ShizCalev** [comment]
Confirming this issue is still active since it's been a while, 3D models (ie characters, treasure chests, jars, ect) still do not render at the proper resolution when increasing internal resolution scale.

(Screenshots are with xemu v0.7.132, at 4x render scaling / 4k. Note the heavy aliasing on the edges of the 3D models, that's what this issue is in reference to since there seemed to be some confusion higher up.)
![2024-09-09_21-47-56-xemu-xemu__v0 7 132](https://github.com/user-attachments/assets/6c646a22-2ba8-4812-9685-788cc4b2758b)
![xemu-2024-09-09-21-50-25](https://github.com/user-attachments/assets/8e987719-7aaf-45ec-a540-cedeb218a91e)
![xemu-2024-09-09-21-52-04](https://github.com/user-attachments/assets/b94adf07-3b1f-4825-b47f-20c558e030d3)
![xemu-2024-09-09-21-57-33](https://github.com/user-attachments/assets/359a5ceb-237a-4664-bbd5-3b21034e95d8)
![xemu-2024-09-09-21-50-18](https://github.com/user-attachments/assets/55c11d55-c615-4dbe-945c-4202c1dd00b2)

**Comment by LDKSuperDante** [comment]
Just commenting to give this a bit of a bump in hopes of it getting more attention, the entire reason I play on Emulators is for playing at higher resolution than on console. I've been playing Warlords on PCSX2 (where this issue isn't present, everything scales properly) but I'm immediately missing Genma's better visual fidelity and new content. Here's to hoping it gets fixed!

**Comment by Triticum0** [comment]
> No, you can keep this open, please just change the title to "Genma Onimusha: 2d backgrounds are not improved by increasing res scale." so it's clear that only the backgrounds are problematic.

@abaire did you still want to do that?

**Comment by Triticum0** [comment]
Also I think is with aliasing no going away from player model when the image when up-scaling which should do looking at other game. Might need hard-scale branch in-order to fixed the issue but need rebase inorder to test it. Same issue with #417 and #433

**Comment by abaire** [comment]
> > No, you can keep this open, please just change the title to "Genma Onimusha: 2d backgrounds are not improved by increasing res scale." so it's clear that only the backgrounds are problematic.
> 
> [@abaire](https://github.com/abaire) did you still want to do that?

No, the screenshots that were added during the discussion show that there is some real issue with the model upscaling here and the current title captures that.

**Comment by SeongGino** [comment]
Yes, the 2D backgrounds are not the issue--it's the 3D models being actively rendered in game not matching the desired scale.

**Comment by Triticum0** [comment]
Here a renderdoc look correct until very last draw call that makes it render at native output res

https://drive.google.com/file/d/1BseIRU0UdLjXMc66wVCk-2r7QtTWLj7m/view?usp=sharing

before

![Image](https://github.com/user-attachments/assets/205173b6-59a5-4837-a77a-ee4fb314b89e)

after

![Image](https://github.com/user-attachments/assets/df159600-bf91-4fac-b54d-201f39167afa)

**Comment by abaire** [comment]
> Here a renderdoc look correct until very last draw call that makes it render at native output res

Very interesting find, nice work!

The last draw (draw 81) is using something that looks like the output of the previous draw (draw 80), but it's a 1x (640x480 pixel) resolution texture that is then stretched to (mostly) fill the 4x sized framebuffer, leading to the crappy visual quality. 

It's especially interesting that draw 80 renders to `0x3DB0000` but the read texture is `0x3120000` and is then rendered on top of `0x3DB0000`, obliterating the higher quality image. The combiner for the last draw just takes everything directly from the texture. I don't see any other draws referring to `0x3120000`, so this could be a CPU blitted buffer (I'm not sure offhand if xemu implements downscaling when reading back from a higher render scale).


Any chance you have the PGRAPH trace that goes with that renderdoc cap? 

If not, I'd be very interested to see a trace that captures several frames so we can see if it's carrying over the texture from the previous frame or if there's some reference that just isn't captured by RenderDoc.

**Comment by ShizCalev** [comment]
I'm curious since those renderdoc images show different stages of lighting between the passes - are the correctly scaled versions from PCSX2 rendered with the darker lighting from the correctly sized version, or the lower-resolution unshaded version from the final draw?

**Comment by Triticum0** [comment]
Here you go

[Genma Onimushares scale PGRAPH.zip](https://github.com/user-attachments/files/22575872/Genma.Onimushares.scale.PGRAPH.zip)

**Comment by abaire** [comment]
> Here you go
> 
> [Genma Onimushares scale PGRAPH.zip](https://github.com/user-attachments/files/22575872/Genma.Onimushares.scale.PGRAPH.zip)

Thanks! Now it makes a bit more sense, `0x312000` is the target of a 2D blit just before it's used as a texture:

```
nv2a_pgraph_method 4: 0x62 -> NV062_SET_COLOR_FORMAT<0x300> (LE_A8R8G8B8)
nv2a_pgraph_method 4: 0x62 -> NV062_SET_PITCH<0x304> (0x0A000A00 {Source:2560, Destination:2560})
nv2a_pgraph_method 4: 0x62 -> NV062_SET_OFFSET_SOURCE<0x308> (0x03DB0000 => 64684032)
nv2a_pgraph_method 4: 0x62 -> NV062_SET_OFFSET_DESTIN<0x30c> (0x03120000 => 51511296)
nv2a_pgraph_method 3: 0x9f -> NV09F_CONTROL_POINT_IN<0x300> (0x00000000 {X:0, Y:0})
nv2a_pgraph_method 3: 0x9f -> NV09F_CONTROL_POINT_OUT<0x304> (0x00000000 {X:0, Y:0})
nv2a_pgraph_method 3: 0x9f -> NV09F_SIZE<0x308> (0x01E00280 {W:640, H:480})
nv2a_pgraph_surface_download [GPU->RAM] COLOR (lin) surface @ 0x03db0000 (w=640,h=480,p=2560,bpp=4)
```

The source is the previous framebuffer. You can see that it uses 640x480 as the size and xemu doesn't currently have a mechanism to realize that the copy will be used as a screen blit, so it is downscaled to be the size expected by the title:

https://github.com/xemu-project/xemu/blob/518208e8494bb0a6189b677fe8f3f693cab41826/hw/xbox/nv2a/pgraph/gl/surface.c#L732

It'll will take some more thought to make sure this can be handled generically, but since the download in this case is triggered by a blit we know that it'll be upscaled as part of the blit and could choose not to downscale. I think the bigger problem would be flagging the fact that the texture is actually <int res scale>x what the game logic expects it to be so it'd need special handling, probably scattered across a few places in the code.

**Comment by ko81e24wy** [comment]
Find something intresting,vulkan backend,2X res makes edge smooth.

**Comment by Zarukyun** [comment]
Still affected to this date.
